File: src/lexer/scanner.rs
//! Scanner for tokenizing Aether source code

use super::token::{Token, TokenKind};
use std::fmt;

/// Lexer error types
#[derive(Debug, Clone, PartialEq)]
pub enum LexerError {
    UnexpectedCharacter(char, usize, usize),
    UnterminatedString(usize, usize),
    InvalidNumber(String, usize, usize),
}

impl fmt::Display for LexerError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LexerError::UnexpectedCharacter(ch, line, col) => {
                write!(f, "Unexpected character '{}' at line {}, column {}", ch, line, col)
            }
            LexerError::UnterminatedString(line, col) => {
                write!(f, "Unterminated string at line {}, column {}", line, col)
            }
            LexerError::InvalidNumber(s, line, col) => {
                write!(f, "Invalid number '{}' at line {}, column {}", s, line, col)
            }
        }
    }
}

impl std::error::Error for LexerError {}

/// Scanner for tokenizing source code
pub struct Scanner {
    source: Vec<char>,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
    column: usize,
}

impl Scanner {
    /// Creates a new scanner from source code
    pub fn new(source: &str) -> Self {
        Self {
            source: source.chars().collect(),
            tokens: Vec::new(),
            start: 0,
            current: 0,
            line: 1,
            column: 1,
        }
    }

    /// Scans all tokens from the source
    pub fn scan_tokens(&mut self) -> Result<Vec<Token>, LexerError> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token()?;
        }

        self.tokens.push(Token::new(
            TokenKind::Eof,
            String::new(),
            self.line,
            self.column,
        ));

        Ok(self.tokens.clone())
    }

    fn scan_token(&mut self) -> Result<(), LexerError> {
        let c = self.advance();
        let start_column = self.column - 1;

        match c {
            ' ' | '\r' | '\t' => {} // Skip whitespace
            '\n' => {
                self.line += 1;
                self.column = 1;
            }
            '(' => self.add_token(TokenKind::LeftParen, start_column),
            ')' => self.add_token(TokenKind::RightParen, start_column),
            '{' => self.add_token(TokenKind::LeftBrace, start_column),
            '}' => self.add_token(TokenKind::RightBrace, start_column),
            '[' => self.add_token(TokenKind::LeftBracket, start_column),
            ']' => self.add_token(TokenKind::RightBracket, start_column),
            ',' => self.add_token(TokenKind::Comma, start_column),
            '.' => self.add_token(TokenKind::Dot, start_column),
            ':' => self.add_token(TokenKind::Colon, start_column),
            '%' => self.add_token(TokenKind::Percent, start_column),
            '+' => {
                let kind = if self.match_char('=') {
                    TokenKind::PlusEqual
                } else {
                    TokenKind::Plus
                };
                self.add_token(kind, start_column);
            }
            '-' => {
                let kind = if self.match_char('=') {
                    TokenKind::MinusEqual
                } else {
                    TokenKind::Minus
                };
                self.add_token(kind, start_column);
            }
            '*' => {
                let kind = if self.match_char('=') {
                    TokenKind::StarEqual
                } else {
                    TokenKind::Star
                };
                self.add_token(kind, start_column);
            }
            '/' => {
                if self.match_char('/') {
                    // Single-line comment
                    while self.peek() != '\n' && !self.is_at_end() {
                        self.advance();
                    }
                } else if self.match_char('*') {
                    // Multi-line comment
                    self.skip_multiline_comment()?;
                } else if self.match_char('=') {
                    self.add_token(TokenKind::SlashEqual, start_column);
                } else {
                    self.add_token(TokenKind::Slash, start_column);
                }
            }
            '!' => {
                let kind = if self.match_char('=') {
                    TokenKind::NotEqual
                } else {
                    TokenKind::Not
                };
                self.add_token(kind, start_column);
            }
            '=' => {
                let kind = if self.match_char('=') {
                    TokenKind::EqualEqual
                } else {
                    TokenKind::Equal
                };
                self.add_token(kind, start_column);
            }
            '<' => {
                let kind = if self.match_char('=') {
                    TokenKind::LessEqual
                } else {
                    TokenKind::Less
                };
                self.add_token(kind, start_column);
            }
            '>' => {
                let kind = if self.match_char('=') {
                    TokenKind::GreaterEqual
                } else {
                    TokenKind::Greater
                };
                self.add_token(kind, start_column);
            }
            '&' => {
                if self.match_char('&') {
                    self.add_token(TokenKind::And, start_column);
                } else {
                    return Err(LexerError::UnexpectedCharacter(c, self.line, start_column));
                }
            }
            '|' => {
                if self.match_char('|') {
                    self.add_token(TokenKind::Or, start_column);
                } else {
                    return Err(LexerError::UnexpectedCharacter(c, self.line, start_column));
                }
            }
            '"' => self.scan_string(start_column)?,
            _ => {
                if c.is_ascii_digit() {
                    self.scan_number(start_column)?;
                } else if c.is_alphabetic() || c == '_' {
                    self.scan_identifier(start_column);
                } else {
                    return Err(LexerError::UnexpectedCharacter(c, self.line, start_column));
                }
            }
        }

        Ok(())
    }

    fn skip_multiline_comment(&mut self) -> Result<(), LexerError> {
        while !self.is_at_end() {
            if self.peek() == '*' && self.peek_next() == '/' {
                self.advance(); // consume '*'
                self.advance(); // consume '/'
                return Ok(());
            }
            if self.peek() == '\n' {
                self.line += 1;
                self.column = 0;
            }
            self.advance();
        }
        Ok(())
    }

    fn scan_string(&mut self, start_column: usize) -> Result<(), LexerError> {
        let mut value = String::new();

        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
                self.column = 0;
            }
            if self.peek() == '\\' {
                self.advance();
                if self.is_at_end() {
                    return Err(LexerError::UnterminatedString(self.line, start_column));
                }
                let escaped = match self.peek() {
                    'n' => '\n',
                    't' => '\t',
                    '\\' => '\\',
                    '"' => '"',
                    _ => self.peek(),
                };
                value.push(escaped);
                self.advance();
            } else {
                value.push(self.advance());
            }
        }

        if self.is_at_end() {
            return Err(LexerError::UnterminatedString(self.line, start_column));
        }

        self.advance(); // closing "

        self.add_token(TokenKind::String(value), start_column);
        Ok(())
    }

    fn scan_number(&mut self, start_column: usize) -> Result<(), LexerError> {
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        let mut is_float = false;
        if self.peek() == '.' && self.peek_next().is_ascii_digit() {
            is_float = true;
            self.advance(); // consume '.'
            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }

        let text: String = self.source[self.start..self.current].iter().collect();

        if is_float {
            match text.parse::<f64>() {
                Ok(value) => self.add_token(TokenKind::Float(value), start_column),
                Err(_) => {
                    return Err(LexerError::InvalidNumber(text, self.line, start_column));
                }
            }
        } else {
            match text.parse::<i64>() {
                Ok(value) => self.add_token(TokenKind::Integer(value), start_column),
                Err(_) => {
                    return Err(LexerError::InvalidNumber(text, self.line, start_column));
                }
            }
        }

        Ok(())
    }

    fn scan_identifier(&mut self, start_column: usize) {
        while self.peek().is_alphanumeric() || self.peek() == '_' {
            self.advance();
        }

        let text: String = self.source[self.start..self.current].iter().collect();
        let kind = match text.as_str() {
            "let" => TokenKind::Let,
            "fn" => TokenKind::Fn,
            "return" => TokenKind::Return,
            "if" => TokenKind::If,
            "else" => TokenKind::Else,
            "while" => TokenKind::While,
            "for" => TokenKind::For,
            "in" => TokenKind::In,
            "break" => TokenKind::Break,
            "continue" => TokenKind::Continue,
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            "null" => TokenKind::Null,
            "import" => TokenKind::Import,
            "from" => TokenKind::From,
            "as" => TokenKind::As,
            _ => TokenKind::Identifier(text.clone()),
        };

        self.add_token(kind, start_column);
    }

    fn add_token(&mut self, kind: TokenKind, column: usize) {
        let text: String = self.source[self.start..self.current].iter().collect();
        self.tokens.push(Token::new(kind, text, self.line, column));
    }

    fn advance(&mut self) -> char {
        let c = self.source[self.current];
        self.current += 1;
        self.column += 1;
        c
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() || self.source[self.current] != expected {
            return false;
        }
        self.current += 1;
        self.column += 1;
        true
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            '\0'
        } else {
            self.source[self.current]
        }
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            '\0'
        } else {
            self.source[self.current + 1]
        }
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }
}

File: src/main/java/com/aether/lexer/Scanner.java
package com.aether.lexer;

import com.aether.exception.LexerException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Lexical analyser for Aether source code.
 *
 * <p>Converts raw source text into a flat list of {@link Token}s terminated by {@link
 * TokenKind#EOF}. Supports:
 *
 * <ul>
 *   <li>Integer and float literals
 *   <li>String literals with escape sequences and {@code ${…}} interpolation
 *   <li>Single-line ({@code //}) and multi-line ({@code /* … * /}) comments
 *   <li>All Aether operators and delimiters including {@code ...} (spread)
 * </ul>
 */
public final class Scanner {

  private static final Map<String, TokenKind> KEYWORDS =
      Map.ofEntries(
          Map.entry("let", TokenKind.LET),
          Map.entry("fn", TokenKind.FN),
          Map.entry("return", TokenKind.RETURN),
          Map.entry("if", TokenKind.IF),
          Map.entry("else", TokenKind.ELSE),
          Map.entry("while", TokenKind.WHILE),
          Map.entry("for", TokenKind.FOR),
          Map.entry("in", TokenKind.IN),
          Map.entry("break", TokenKind.BREAK),
          Map.entry("continue", TokenKind.CONTINUE),
          Map.entry("true", TokenKind.TRUE),
          Map.entry("false", TokenKind.FALSE),
          Map.entry("null", TokenKind.NULL),
          Map.entry("import", TokenKind.IMPORT),
          Map.entry("from", TokenKind.FROM),
          Map.entry("as", TokenKind.AS),
          Map.entry("try", TokenKind.TRY),
          Map.entry("catch", TokenKind.CATCH),
          Map.entry("throw", TokenKind.THROW),
          Map.entry("struct", TokenKind.STRUCT));

  private final String source;
  private final List<Token> tokens = new ArrayList<>();
  private int start = 0;
  private int current = 0;
  private int line = 1;
  private int column = 1;

  public Scanner(String source) {
    this.source = source;
  }

  /**
   * Scans all tokens from the source and returns the resulting list.
   *
   * @return immutable list of tokens ending with EOF
   * @throws LexerException on unexpected characters or unterminated strings
   */
  public List<Token> scanTokens() {
    while (!isAtEnd()) {
      start = current;
      scanToken();
    }
    tokens.add(Token.of(TokenKind.EOF, "", line, column));
    return List.copyOf(tokens);
  }

  // ── Private scan methods ─────────────────────────────────────────────────────

  private void scanToken() {
    char c = advance();
    int startCol = column - 1;

    switch (c) {
      case ' ', '\r', '\t' -> {} // skip whitespace
      case '\n' -> {} // line/column already updated in advance()
      case '(' -> addTokenAt(TokenKind.LEFT_PAREN, startCol);
      case ')' -> addTokenAt(TokenKind.RIGHT_PAREN, startCol);
      case '{' -> addTokenAt(TokenKind.LEFT_BRACE, startCol);
      case '}' -> addTokenAt(TokenKind.RIGHT_BRACE, startCol);
      case '[' -> addTokenAt(TokenKind.LEFT_BRACKET, startCol);
      case ']' -> addTokenAt(TokenKind.RIGHT_BRACKET, startCol);
      case ',' -> addTokenAt(TokenKind.COMMA, startCol);
      case ':' -> addTokenAt(TokenKind.COLON, startCol);
      case '%' -> addTokenAt(TokenKind.PERCENT, startCol);
      case '.' -> {
        if (peek() == '.' && peekNext() == '.') {
          advance();
          advance();
          addTokenAt(TokenKind.SPREAD, startCol);
        } else {
          addTokenAt(TokenKind.DOT, startCol);
        }
      }
      case '+' -> addTokenAt(matchChar('=') ? TokenKind.PLUS_EQUAL : TokenKind.PLUS, startCol);
      case '-' -> addTokenAt(matchChar('=') ? TokenKind.MINUS_EQUAL : TokenKind.MINUS, startCol);
      case '*' -> addTokenAt(matchChar('=') ? TokenKind.STAR_EQUAL : TokenKind.STAR, startCol);
      case '/' -> {
        if (matchChar('/')) {
          skipLineComment();
        } else if (matchChar('*')) {
          skipBlockComment();
        } else if (matchChar('=')) {
          addTokenAt(TokenKind.SLASH_EQUAL, startCol);
        } else {
          addTokenAt(TokenKind.SLASH, startCol);
        }
      }
      case '!' -> addTokenAt(matchChar('=') ? TokenKind.NOT_EQUAL : TokenKind.NOT, startCol);
      case '=' -> addTokenAt(matchChar('=') ? TokenKind.EQUAL_EQUAL : TokenKind.EQUAL, startCol);
      case '<' -> addTokenAt(matchChar('=') ? TokenKind.LESS_EQUAL : TokenKind.LESS, startCol);
      case '>' ->
          addTokenAt(matchChar('=') ? TokenKind.GREATER_EQUAL : TokenKind.GREATER, startCol);
      case '&' -> {
        if (matchChar('&')) {
          addTokenAt(TokenKind.AND, startCol);
        } else {
          throw new LexerException("Unexpected character '&'", line, startCol);
        }
      }
      case '|' -> {
        if (matchChar('|')) {
          addTokenAt(TokenKind.OR, startCol);
        } else {
          throw new LexerException("Unexpected character '|'", line, startCol);
        }
      }
      case '"' -> scanString(startCol);
      default -> {
        if (Character.isDigit(c)) {
          scanNumber(startCol);
        } else if (Character.isLetter(c) || c == '_') {
          scanIdentifier(startCol);
        } else {
          throw new LexerException("Unexpected character '" + c + "'", line, startCol);
        }
      }
    }
  }

  private void skipLineComment() {
    while (peek() != '\n' && !isAtEnd()) {
      advance();
    }
  }

  private void skipBlockComment() {
    while (!isAtEnd()) {
      if (peek() == '*' && peekNext() == '/') {
        advance(); // consume '*'
        advance(); // consume '/'
        return;
      }
      advance();
    }
  }

  private void scanString(int startCol) {
    StringBuilder currentLiteral = new StringBuilder();
    List<StringPart> parts = new ArrayList<>();
    boolean hasInterpolation = false;

    while (peek() != '"' && !isAtEnd()) {
      if (peek() == '$' && peekNext() == '{') {
        hasInterpolation = true;
        advance(); // consume '$'
        advance(); // consume '{'

        parts.add(new StringPart.Literal(currentLiteral.toString()));
        currentLiteral.setLength(0);

        // Collect expression source until matching '}'
        StringBuilder exprSrc = new StringBuilder();
        int depth = 1;
        while (!isAtEnd() && depth > 0) {
          char ch = advance();
          if (ch == '{') {
            depth++;
            exprSrc.append(ch);
          } else if (ch == '}') {
            depth--;
            if (depth > 0) {
              exprSrc.append(ch);
            }
          } else {
            exprSrc.append(ch);
          }
        }
        parts.add(new StringPart.Placeholder(exprSrc.toString()));
      } else if (peek() == '\\') {
        advance(); // consume backslash
        if (isAtEnd()) {
          throw new LexerException("Unterminated string", line, startCol);
        }
        char escaped =
            switch (peek()) {
              case 'n' -> '\n';
              case 't' -> '\t';
              case '\\' -> '\\';
              case '"' -> '"';
              default -> peek();
            };
        currentLiteral.append(escaped);
        advance();
      } else {
        currentLiteral.append(advance());
      }
    }

    if (isAtEnd()) {
      throw new LexerException("Unterminated string", line, startCol);
    }
    advance(); // closing "

    if (hasInterpolation) {
      parts.add(new StringPart.Literal(currentLiteral.toString()));
      tokens.add(Token.ofStringInterp(parts, source.substring(start, current), line, startCol));
    } else {
      tokens.add(
          Token.ofString(currentLiteral.toString(), source.substring(start, current), line, startCol));
    }
  }

  private void scanNumber(int startCol) {
    while (Character.isDigit(peek())) {
      advance();
    }

    boolean isFloat = peek() == '.' && Character.isDigit(peekNext());
    if (isFloat) {
      advance(); // consume '.'
      while (Character.isDigit(peek())) {
        advance();
      }
    }

    String text = source.substring(start, current);
    if (isFloat) {
      tokens.add(Token.ofFloat(Double.parseDouble(text), text, line, startCol));
    } else {
      tokens.add(Token.ofInt(Long.parseLong(text), text, line, startCol));
    }
  }

  private void scanIdentifier(int startCol) {
    while (Character.isLetterOrDigit(peek()) || peek() == '_') {
      advance();
    }

    String text = source.substring(start, current);
    TokenKind kind = KEYWORDS.getOrDefault(text, TokenKind.IDENTIFIER);
    tokens.add(Token.of(kind, text, line, startCol));
  }

  // ── Helpers ──────────────────────────────────────────────────────────────────

  private boolean isAtEnd() {
    return current >= source.length();
  }

  private char advance() {
    char c = source.charAt(current++);
    if (c == '\n') {
      line++;
      column = 1;
    } else {
      column++;
    }
    return c;
  }

  private char peek() {
    return isAtEnd() ? '\0' : source.charAt(current);
  }

  private char peekNext() {
    return current + 1 >= source.length() ? '\0' : source.charAt(current + 1);
  }

  private boolean matchChar(char expected) {
    if (isAtEnd() || source.charAt(current) != expected) {
      return false;
    }
    advance();
    return true;
  }

  private void addToken(TokenKind kind) {
    String lexeme = source.substring(start, current);
    tokens.add(Token.of(kind, lexeme, line, column - lexeme.length()));
  }

  private void addTokenAt(TokenKind kind, int col) {
    String lexeme = source.substring(start, current);
    tokens.add(Token.of(kind, lexeme, line, col));
  }
}
